pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -206,16 +207,22 @@ func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconci
 		return r.reconcileAndError(instance, err)
 	}
 
-	if err = runCordonOrUncordon(r, node, true); err != nil {
+	if err = drain.RunCordonOrUncordon(r.drainer, node, true); err != nil {
 		return r.reconcileAndError(instance, err)
 	}
 
-	stop := make(chan struct{})
-	defer close(stop)
+	podsToDelete, errors := r.drainer.GetPodsForDeletion(nodeName)
 
-	reqLogger.Infof("Evict all Pods from Node: %s", nodeName)
-	if err = drainPods(r, node, stop); err != nil {
-		return r.reconcileAndError(instance, err)
+	if errors != nil && len(errors) != 0 {
+		aerror := utilerrors.NewAggregate(errors)
+		reqLogger.Errorf("Errors while listing pods for deletion: %v", aerror)
+	}
+
+	if len(podsToDelete.Pods()) != 0 {
```

Comment:
since you are not reconciling again if `GetPodsForDeletion` returns an error can you end up with podsToDelete being nil ? if so then you will get an unexpected error here with calling Pods() on nil.

Reply:
changed in a later commit. please review.